### ISSUE TITLE
Feature: discard operator for terminals?

### ISSUE THREAD

**Issue by mindplay-dk** [jcoglan/canopy#15]
Played around with this on friday, and the biggest distraction while trying to develop a grammar, was whitespace and comment nodes, which generate a lot of noise.

It's pretty common to need to allow optional whitespace (and/or comments) in or around various statements and blocks, and I suppose there could be other types of terminals (commas or other separators, string delimiters, etc.) which could be usefully discarded.

How about adding a discard operator?

Maybe something like a prefix operator `@` to "mute" an expression - here's a small portion of JSON grammar with whitespace and commas muted:

```
grammar CanopyJson

document  <-  @__ (object / array) @__

object    <-  "{" pair (@"," pair)* "}" / "{" @__ "}"

...

__        <-  [\s]*
```

Would that work?

**Comment by jcoglan** [comment]
Apologies for the delay on this, I've been busy with other projects.

I've committed this, which I believe adds the behaviour you're after: https://github.com/jcoglan/canopy/commit/e49fa5c486a82fdb50e0470e3d155142d0a5f0bd. For expressions that are part of a sequence, you can prefix them with `@` and they will be excluded from the elements of the sequence in the parse tree, and from the list of elements passed to `%action` functions.

Does that match what you were expecting?

**Comment by mindplay-dk** [comment]
That's exactly what I was looking for :-)

**Comment by jcoglan** [comment]
Great :) I'm in the process of tidying up the grammar and I'll try to get a release done soon.

**Comment by guioconnor** [comment]
Hello.

I believe this never made it to production; npm's latest release date is 3y ago and a quick try breaks the compilation on the `@` symbol. Is there any plan to release this feature?

**Comment by y8** [comment]
Could you please make a release with this feature? This is really hard do deal with whitespace/comments nodes

**Comment by jcoglan** [comment]
This should now be fixed as part of the v0.4 release; see https://blog.jcoglan.com/2022/04/01/announcing-canopy-0-4/
